gcp/volume.go

```diff
@@ -158,30 +199,59 @@ func (c *Client) getVolumeByID(volume volumeRequest) (volumeResult, error) {
 		log.Print("Failed to unmarshall response from getVolumeByID")
 		return volumeResult{}, err
 	}
+
+	// volumeID is verified by Terraform. If we use ID = <volumeID>:<region>, we need to revert our ID changes
+	if originalID != "" {
+		result.VolumeID = originalID
+	}
 	return result, nil
 }
 
-func (c *Client) getVolumeByRegion(region string) ([]volumeResult, error) {
+// refactored, but commented, since no code is using it currently
+// func (c *Client) getVolumeByRegion(region string) ([]volumeResult, error) {
+// 	return c.getVolumes(region)
+// }
+
+// Returns volumes of the project. region = "-" for all regions
+func (c *Client) getVolumes(region string) ([]volumeResult, error) {
 
 	baseURL := fmt.Sprintf("%s/Volumes", region)
-	var volumes []volumeResult
+	var result []volumeResult
 
 	statusCode, response, err := c.CallAPIMethod("GET", baseURL, nil)
 	if err != nil {
-		log.Print("ListVolumes request failed")
-		return volumes, err
+		log.Print("getVolumes request failed")
+		return result, err
 	}
 
-	responseError := apiResponseChecker(statusCode, response, "getVolumeByRegion")
+	responseError := apiResponseChecker(statusCode, response, "getVolumes")
 	if responseError != nil {
-		return volumes, responseError
+		return result, responseError
 	}
 
-	if err := json.Unmarshal(response, &volumes); err != nil {
-		log.Print("Failed to unmarshall response from getVolumeByRegion")
-		return volumes, err
+	if err := json.Unmarshal(response, &result); err != nil {
+		log.Print("Failed to unmarshall response from getVolumes")
+		return result, err
+	}
+	return result, nil
+}
+
+// Filter all volumes of the project by applying a filter function
+// Example filter function: func(v volumeResult) bool { return v.VolumeID == "1234-5678-90" }
+func (c *Client) filterAllVolumes(f func(volumeResult) bool) ([]volumeResult, error) {
+	filteredVolumes := make([]volumeResult, 0)
+
+	vols, err := c.getVolumes("-")
```

Comment:
Are we getting all volumes in all regions?    This sounds like an expensive operation.

Reply:
This is a helper function for the use cases where we need to get a list or sublist of all volumes. It is used only in the specific cases where this is necessary.
In our code, we have two places where this is needed:
1) getVolumeByRegion, which isn't used anywhere in the code so I commented it out
2) getVolumeByID. Here it is only required in the import case, where the user specified a volume UUID, but no region. We use it to lookup the specified UUID in all regions.